src/shared/wall.c

```diff
@@ -34,96 +34,95 @@ static int write_to_terminal(const char *tty, const char *message) {
                                 return -ENOMEM;
-                        path = buf;
+
+                        tty_path = p;
                 }
 
-                /* It seems that the address field is always set for remote logins.
-                 * For local logins and other local entries, we get [0,0,0,0]. */
-                bool is_local = memeqzero(u->ut_addr_v6, sizeof(u->ut_addr_v6));
+                /* It seems that the address field is always set for remote logins. For local logins and
+                 * other local entries, we get [0,0,0,0]. */
+                is_local = eqzero(u->ut_addr_v6);
 
-                if (!match_tty || match_tty(path, is_local, userdata)) {
-                        q = write_to_terminal(path, message);
-                        if (q < 0)
-                                r = q;
-                }
+                if (!match_tty || match_tty(tty_path, is_local, userdata))
+                        RET_GATHER(r, write_to_terminal(tty_path, message));
         }
 
         return r;
 }
+#endif
 
-#else
-
-static int do_wall(
+#if ENABLE_LOGIND
+_unused_ static int wall_logind(
         const char *message,
-        const char *username,
-        const char *origin_tty,
         bool (*match_tty)(const char *tty, bool is_local, void *userdata),
         void *userdata) {
 
-        int r;
         _cleanup_strv_free_ char **sessions = NULL;
+        int r;
+
+        assert(message);
 
         r = sd_get_sessions(&sessions);
-        if (r < 0)
+        if (r <= 0)
                 return r;
 
+        r = 0;
+
         STRV_FOREACH(s, sessions) {
-                _cleanup_free_ char *path = NULL, *tty = NULL, *rhost = NULL;
+                _cleanup_free_ char *tty_path = NULL, *tty = NULL, *rhost = NULL;
+                bool is_local;
                 int q;
 
                 q = sd_session_get_tty(*s, &tty);
-                if (q < 0) {
-                        if (q != -ENXIO && q != -ENODATA)
-                                r = q;
+                if (IN_SET(q, -ENXIO, -ENODATA))
                         continue;
-                }
+                if (q < 0)
+                        return RET_GATHER(r, q);
 
-                path = strjoin("/dev/", tty);
-                if (!path)
+                tty_path = strjoin("/dev/", tty);
+                if (!tty_path)
                         return -ENOMEM;
 
                 (void) sd_session_get_remote_host(*s, &rhost);
-                bool is_local = !rhost;
+                is_local = !rhost;
 
-                if (!match_tty || match_tty(path, is_local, userdata)) {
-                        q = write_to_terminal(path, message);
-                        if (q < 0)
-                                r = q;
-                }
+                if (!match_tty || match_tty(tty_path, is_local, userdata))
+                        RET_GATHER(r, write_to_terminal(tty_path, message));
         }
+
         return r;
 }
-
 #endif
 
 int wall(
```

Comment:
#if ENABLE_UTMP || ENABLE_LOGIND missing here, otherwise you'll define botht the static inline and this version of the function in that case.